lib/profileAgents.js

```diff
@@ -0,0 +1,356 @@
+const {promisify} = require('util');
+const kms = require('./kms');
+const utils = require('./utils');
+const zcaps = require('./zcaps');
+
+const {util: {BedrockError}} = bedrock;
+
+// load config defaults
+require('./config');
+const KMS_MODULE = 'ssm-v1';
+
+bedrock.events.on('bedrock-mongodb.ready', async () => {
+  await promisify(database.openCollections)(['profile-profileAgent']);
+
+  await promisify(database.createIndexes)([{
+    // cover queries of a profile agent by its ID
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.id': 1},
+    options: {unique: true, background: false}
+  }, {
+    // cover queries of a profile agent by its ID
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.profile': 1},
+    options: {unique: false, background: false}
+  }, {
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.controller': 1},
+    options: {unique: true, background: false}
+  }, {
+    // cover queries of a profile agent by an account ID
+    collection: 'profile-profileAgent',
+    fields: {'profileAgent.account': 1, 'profileAgent.id': 1},
+    options: {
+      partialFilterExpression: {account: {$exists: true}},
+      unique: true,
+      background: false
+    }
+  }]);
+});
+
+/**
+ * Creates a ProfileAgent and associates it with an account if provided.
+ *
+ * @param {Object} account the account to associate with the ProfileAgent.
+ *
+ * @return {Promise<ProfileAgent>} resolves to a ProfileAgent.
+ */
+exports.create = async ({account}) => {
+  // 1. Generate a random secret
+  const secret = base64url.encode(crypto.randomBytes(32));
+  const handle = 'primary';
+  // 2. Generate capability agent for the zCap key
+  const capabilityAgent = await CapabilityAgent.fromSecret({handle, secret});
+  // 3. Create keystore in order to create the zCap key for the profile agent
+  // TODO: Will want to store the capability agent up here first, otherwise we
+  //       have no clean way of cleaning up the other stuff that gets created
+  //       on other systems should we experience a failure.
+  const keystore = await kms.createKeystore(
+    {capabilityAgent, referenceId: 'primary'});
+  // 4. Create the zCap key for the profile agent
+  const keystoreAgent = kms.getKeystoreAgent({capabilityAgent, keystore});
+  const key = await keystoreAgent.generateKey(
+    {type: 'Ed25519VerificationKey2018', kmsModule: KMS_MODULE});
+  // 5. Use the zCap key to create the DID for the profile agent
+  await utils.setKeyId({key});
+  // 6. Generate capability agent for ProfileAgent
+  const paZcapAgent = new CapabilityAgent({handle: 'primary', signer: key});
+  // 7. Create keystore to store keys controlled by the profile agent
+  const paKeystore = await kms.createKeystore(
+    {capabilityAgent: paZcapAgent, referenceId: 'primary'});
+  const paKeystoreAgent = kms.getKeystoreAgent(
+    {capabilityAgent: paZcapAgent, keystore: paKeystore});
+
+  const did = key.id.split('#')[0];
```

Comment:
Is this the algo we're going with for this, there was some discussion about helper APIs for this sort of thing.

Reply:
This all I have. Don't know if others have another one we should be using.

@dlongley @dmitrizagidulin

Reply:
```suggestion
  const did = paZcapAgent.id;
```

The above should work for now. We'll want to clean that up in `CapabilityAgent` because it's making the same assumption, but we can handle it over there.